### PR TITLE
Add storage driver context

### PR DESCRIPTION
**Pull request by amitshukla**
Errors thrown by storage drivers don't have the name of the driver, causing user confusion about whether the error is coming from Docker or from a storage driver. This change adds the storage driver name to each error message.

This required changing ErrUnsupportedDriver to a type, leading to code changes whenever ErrUnsupportedDriver is used.
The tests check whether the driver name appears in the error message.

Closes: #664 

Signed-off-by: Amit Shukla amit.shukla@docker.com

### CODE REVIEW

**Comment by RichardScothern** [comment]
You will need to sign the commit correctly (you are missing `<` and `>` around your email address)

**Comment by amitshukla** [comment]
@RichardScothern done - thanks

File: registry/storage/blobserver.go [hunk 1]
```diff
@@ -36,16 +36,15 @@ func (bs *blobServer) ServeBlob(ctx context.Context, w http.ResponseWriter, r *h
 
 	redirectURL, err := bs.driver.URLFor(ctx, path, map[string]interface{}{"method": r.Method})
 
-	switch err {
-	case nil:
+	if err == nil {
```
Comment (RichardScothern) [inline registry/storage/blobserver.go]: Was this change necessary?

File: registry/storage/driver/base/base.go [hunk 2]
```diff
@@ -50,16 +50,40 @@ type Base struct {
 	storagedriver.StorageDriver
 }
 
+// Format errors received from the storage driver
+func (base *Base) setDriverName(e error) error {
+	if e != nil {
```
Comment (RichardScothern) [inline registry/storage/driver/base/base.go]: go conventions are to return early and leave code as least indented as possible.  Could you invert this if statement?

File: registry/storage/driver/filesystem/driver.go [hunk 3]
```diff
@@ -248,7 +248,7 @@ func (d *driver) Delete(ctx context.Context, subPath string) error {
 // URLFor returns a URL which may be used to retrieve the content stored at the given path.
 // May return an UnsupportedMethodErr in certain StorageDriver implementations.
 func (d *driver) URLFor(ctx context.Context, path string, options map[string]interface{}) (string, error) {
-	return "", storagedriver.ErrUnsupportedMethod
+	return "", new(storagedriver.ErrUnsupportedMethod)
```
Comment (RichardScothern) [inline registry/storage/driver/filesystem/driver.go]: what is the motivation for `new`ing these?

File: registry/storage/driver/base/base.go [hunk 4]
```diff
@@ -50,16 +50,40 @@ type Base struct {
...
+		if actualErr, ok := e.(storagedriver.ErrUnsupportedMethod); ok {
```
Comment (RichardScothern) [inline registry/storage/driver/base/base.go]: It might be more extensible to use `reflect` to check if the struct has a `DriverName` field and set it appropriately.

**Comment by dmp42** [comment]
@amitshukla heads-up! release is near :)

File: registry/storage/blobserver.go [hunk 5]
```diff
@@ -36,16 +36,15 @@ func (bs *blobServer) ServeBlob(ctx context.Context, w http.ResponseWriter, r *h
...
 		if bs.redirect {
 			// Redirect to storage URL.
 			http.Redirect(w, r, redirectURL, http.StatusTemporaryRedirect)
 			return err
 		}
+	}
 
-		fallthrough
-	case driver.ErrUnsupportedMethod:
+	if _, ok := err.(*driver.ErrUnsupportedMethod); ok {
```
Comment (stevvooe) [inline registry/storage/blobserver.go]: No reason to use pointer type for this struct.

File: registry/storage/driver/base/base.go (hunk 4 above)
Comment (stevvooe) [inline registry/storage/driver/base/base.go]: Please use a type switch here.

File: registry/storage/driver/base/base.go [hunk 6]
```diff
@@ -50,16 +50,40 @@ type Base struct {
 	storagedriver.StorageDriver
 }
 
+// Format errors received from the storage driver
+func (base *Base) setDriverName(e error) error {
```
Comment (stevvooe) [inline registry/storage/driver/base/base.go]: Lose this method and just set the correct value at the error generation site.

File: registry/storage/driver/filesystem/driver.go (hunk 3 above)
Comment (amitshukla) [inline registry/storage/driver/filesystem/driver.go]: I changed ErrUnsupportedMethod from var to type. As a result I needed to new the type to return an object of the appropriate type.

File: registry/storage/blobserver.go (hunk 1 above)
Comment (amitshukla) [inline registry/storage/blobserver.go]: I changed ErrUnsupportedMethod from var to type. This meant switch on err doesn't work any more.

Comment (stevvooe) [inline registry/storage/blobserver.go]: @amitshukla That is untrue. You just need to use a type switch. :)

File: registry/storage/driver/filesystem/driver.go (hunk 3 above)
Comment (stevvooe) [inline registry/storage/driver/filesystem/driver.go]: You can return a concrete value with the syntax `storagedriver.ErrUnsupportedMethod{}`. I don't recommend using pointer types for errors.

**Comment by RichardScothern** [comment]
ping @amitshukla

**Comment by RichardScothern** [comment]
carried in #1150
